diff --git a/gfx/cairo/cairo/src/cairo-array.c b/gfx/cairo/cairo/src/cairo-array.c
--- a/gfx/cairo/cairo/src/cairo-array.c
+++ b/gfx/cairo/cairo/src/cairo-array.c
@@ -181,6 +181,9 @@ _cairo_array_index (cairo_array_t *array, unsigned int index)

     assert (index < array->num_elements);

+    if (unlikely (index >= array->num_elements))
+	abort ();
+
     return array->elements + index * array->element_size;
 }

@@ -225,6 +228,9 @@ _cairo_array_index_const (const cairo_array_t *array, unsigned int index)

     assert (index < array->num_elements);

+    if (unlikely (index >= array->num_elements))
+	abort ();
+
     return array->elements + index * array->element_size;
 }
